File: ui/macos/Sources/VibeSync/Net/Protocol.swift
// Protocol.swift — frontière Swift ↔ protocole C commun (ADR-010, phase 4).
//
// L'encodage et le décodage de docs/protocol.md n'existent plus qu'en un
// exemplaire pour les deux clients natifs : `core/src/protocol.c`, au-dessus du
// parseur `core/src/json.c`. Ce fichier ne décide de RIEN — il traduit le
// vocabulaire Swift (ClientMessage, Welcome, RoomState…) en appels C et
// retraduit ce qui en sort. Toute règle de lecture qui se trouverait ici serait
// un bug de conception : elle divergerait du client Windows le jour même.
//
// Ce que le C apporte et que le décodage Swift n'avait pas (docs/protocol.md
// §forward-compat) : un message dont un champ OBLIGATOIRE manque ou est mal
// typé est INVALIDE, donc ignoré, au lieu d'être rempli de zéros silencieux —
// un `pong` vide devenait {t:0, serverMs:0} et empoisonnait l'offset d'horloge.
// C'est déjà ce que fait `on_server_message` dans ui/win32/src/main.c.

import Foundation
import VSCore

public let protocolVersion: Int64 = Int64(VS_PROTOCOL_VERSION)
/// Jeton de reprise de session : 16 octets aléatoires en hexadécimal.
public let sessionTokenBytes: Int = Int(VS_SESSION_TOKEN_BYTES)
public let sessionTokenLength: Int = Int(VS_SESSION_TOKEN_LEN)

// MARK: - Serveur → client

public struct ServerUser {
    public var id: String = ""
    public var name: String = ""
    public var ready: Bool = false
    public var positionSec: Double = 0
    public var latencyMs: Int64 = 0
    public var hasFile: Bool = false
    public var fileName: String = ""
    public var fileDurationSec: Double = 0
    public var fileSizeBytes: Int64 = 0

    public init() {}
}

/// Contenu du `welcome` (docs/protocol.md §Messages serveur → client). Un
/// struct plutôt qu'un n-uplet : le message a grossi avec VS-023 et chaque
/// champ mérite son nom.
public struct Welcome {
    public var selfId: String = ""
    public var room: String = ""
    /// Optionnel côté Swift par héritage ; en pratique le C n'accepte pas un
    /// welcome sans état de salle recevable (il l'invalide), donc ce champ est
    /// toujours renseigné quand un welcome arrive jusqu'ici.
    public var state: RoomState? = nil
    public var users: [ServerUser] = []
    /// Ready que le serveur nous attribue. Le moteur ne l'adopte PAS au welcome
    /// (c'est notre état local qui fait foi au re-join) : ce champ n'est là que
    /// pour l'affichage et le diagnostic.
    public var selfReady: Bool? = nil
    /// VS-023 : version applicative du serveur et lien de téléchargement.
    /// Champs additifs, absents des serveurs plus anciens.
    public var serverVersion: String = ""
    public var downloadUrl: String = ""

    public init() {}
}

public enum ServerMessage {
    case welcome(Welcome)
    case pong(Pong)
    case roomState(RoomState)
    case users([ServerUser])
    case chatEvent(from: String, text: String, serverMs: Int64)
    case toast(level: String, text: String)
    case error(code: String, text: String)
    /// Forward-compat : type inconnu, à ignorer sans fermer la connexion.
    case unknown(type: String)
}

public enum Proto {

    // MARK: Client → serveur

    /// Premier message de la session. `password` et `session` sont omis s'ils
    /// sont vides (proto_encode_hello).
    public static func encodeHello(name: String,
                                   room: String,
                                   password: String,
                                   session: String) -> String {
        return Scratch.shared.use { arena in
            withStr8(name) { n in
                withStr8(room) { r in
                    withStr8(password) { p in
                        withStr8(session) { s in
                            coreString(proto_encode_hello(arena, n, r, p, s))
                        }
                    }
                }
            }
        }
    }

    /// Encode une décision du moteur (proto_encode_msg).
    public static func encode(_ m: ClientMessage) -> String {
        var msg = VsMsg()
        switch m {
        case .ping(let t):
            msg.kind = VS_MSG_PING
            msg.t = t
        case .setReady(let ready):
            msg.kind = VS_MSG_SET_READY
            msg.ready = ready ? 1 : 0
        case .setFile(let name, let durationSec, let sizeBytes):
            msg.kind = VS_MSG_SET_FILE
            withStr8(name) { strbuf_set(&msg.name, $0) }
            msg.duration_sec = durationSec
            msg.size_bytes = sizeBytes
        case .control(let action, let positionSec):
            msg.kind = VS_MSG_CONTROL
            msg.action = Proto.vsAction(action)
            msg.position_sec = positionSec
        case .report(let positionSec, let paused, let buffering):
            msg.kind = VS_MSG_REPORT
            msg.position_sec = positionSec
            msg.paused = paused ? 1 : 0
            msg.buffering = buffering ? 1 : 0
        case .chat(let text):
            msg.kind = VS_MSG_CHAT
            withStr8(text) { strbuf_set(&msg.text, $0) }
        }
        return Scratch.shared.use { arena in
            coreString(proto_encode_msg(arena, &msg))
        }
    }

    private static func vsAction(_ a: ControlAction) -> VsAction {
        switch a {
        case .play: return VS_ACT_PLAY
        case .pause: return VS_ACT_PAUSE
        case .seek: return VS_ACT_SEEK
        }
    }

    // MARK: Serveur → client

    /// Analyse une enveloppe {type, data}. `nil` si le message est illisible
    /// (JSON invalide, enveloppe sans type) **ou invalide** (type connu mais
    /// champ obligatoire absent ou mal typé) : dans les deux cas il est ignoré
    /// sans fermer la connexion, comme le fait le client Windows.
    public static func decode(_ raw: String) -> ServerMessage? {
        return Scratch.shared.use { arena -> ServerMessage? in
            withStr8(raw) { text -> ServerMessage? in
                guard let m = proto_decode(arena, text), m.pointee.invalid == 0 else {
                    return nil
                }
                // Tout est recopié ici : les Str8 du message pointent dans
                // l'arène, qui est rendue à la sortie de `use`.
                return message(m.pointee)
            }
        }
    }

    private static func message(_ m: VsInMsg) -> ServerMessage {
        switch m.kind {
        case VS_IN_WELCOME:
            var w = Welcome()
            w.selfId = coreString(m.self_id)
            w.room = coreString(m.room)
            if m.have_state != 0 {
                w.state = roomState(m.state)
            }
            w.users = users(m)
            if m.have_self_ready != 0 {
                w.selfReady = m.self_ready != 0
            }
            w.serverVersion = coreString(m.server_version)
            w.downloadUrl = coreString(m.download_url)
            return .welcome(w)

        case VS_IN_PONG:
            return .pong(Pong(t: m.pong.t, serverMs: m.pong.server_ms))

        case VS_IN_ROOMSTATE:
            return .roomState(roomState(m.state))

        case VS_IN_USERS:
            return .users(users(m))

        case VS_IN_CHATEVENT:
            return .chatEvent(from: coreString(m.from),
                              text: coreString(m.text),
                              serverMs: m.server_ms)

        case VS_IN_TOAST:
            return .toast(level: coreString(m.level), text: coreString(m.text))

        case VS_IN_ERROR:
            return .error(code: coreString(m.code), text: coreString(m.text))

        default:
            return .unknown(type: coreString(m.type))
        }
    }

    private static func roomState(_ rs: VsRoomState) -> RoomState {
        return RoomState(paused: rs.paused != 0,
                         positionSec: rs.position_sec,
                         rate: rs.rate,
                         refServerMs: rs.ref_server_ms,
                         setBy: coreString(rs.set_by))
    }

    private static func users(_ m: VsInMsg) -> [ServerUser] {
        guard let raw = m.users, m.user_count > 0 else {
            return []
        }
        var out: [ServerUser] = []
        out.reserveCapacity(Int(m.user_count))
        for index in 0..<Int(m.user_count) {
            let u = raw[index]
            var user = ServerUser()
            user.id = coreString(u.id)
            user.name = coreString(u.name)
            user.ready = u.ready != 0
            user.positionSec = u.position_sec
            user.latencyMs = u.latency_ms
            user.hasFile = u.has_file != 0
            user.fileName = coreString(u.file_name)
            user.fileDurationSec = u.file_duration_sec
            user.fileSizeBytes = u.file_size_bytes
            out.append(user)
        }
        return out
    }

    /// Un code d'erreur fatal interdit de réessayer (proto_error_is_fatal).
    public static func isFatal(_ code: String) -> Bool {
        return withStr8(code) { proto_error_is_fatal($0) != 0 }
    }

    // MARK: Jeton de session

    /// 16 octets du générateur du système, en hexadécimal minuscule
    /// (proto_session_token → vs_random_bytes → arc4random_buf).
    public static func sessionToken() -> String {
        var buffer = [CChar](repeating: 0, count: sessionTokenLength + 1)
        let ok = buffer.withUnsafeMutableBufferPointer { raw in
            proto_session_token(raw.baseAddress, Int(raw.count))
        }
        if ok == 0 {
            // Le générateur du système n'échoue pas en pratique ; s'il le
            // faisait, un jeton non aléatoire vaut mieux qu'un jeton vide (il
            // ne sert qu'à reconnaître notre propre session zombie).
            var bytes = [UInt8](repeating: 0, count: sessionTokenBytes)
            for i in 0..<bytes.count {
                bytes[i] = UInt8.random(in: 0...255)
            }
            return hex(bytes)
        }
        return String(cString: buffer)
    }

    /// Forme attendue d'un jeton relu des réglages (proto_session_token_valid).
    public static func isValidSessionToken(_ token: String) -> Bool {
        return withStr8(token) { proto_session_token_valid($0) != 0 }
    }

    /// Hexadécimal minuscule (proto_hex).
    public static func hex(_ bytes: [UInt8]) -> String {
        if bytes.isEmpty {
            return ""
        }
        var out = [CChar](repeating: 0, count: bytes.count * 2 + 1)
        out.withUnsafeMutableBufferPointer { raw in
            proto_hex(bytes, bytes.count, raw.baseAddress)
        }
        return String(cString: out)
    }
}

File: ui/win32/src/protocol.c
#include "protocol.h"

#include <string.h>

// --------------------------------------------------------------- encodage ---

// env_begin ouvre l'enveloppe {"type":…,"data":{…}}.
static void env_begin(JsonWriter *w, const char *type) {
    jw_obj_begin(w);
    jw_key(w, "type");
    jw_cstr(w, type);
    jw_key(w, "data");
    jw_obj_begin(w);
}

static Str8 env_end(JsonWriter *w) {
    jw_obj_end(w);  // data
    jw_obj_end(w);  // enveloppe
    return jw_result(w);
}

Str8 proto_encode_hello(Arena *a, Str8 name, Str8 room, Str8 password, Str8 session) {
    JsonWriter w;
    jw_init(&w, a);
    env_begin(&w, "hello");
    jw_kv_i64(&w, "version", VS_PROTOCOL_VERSION);
    jw_kv_str(&w, "name", name);
    jw_kv_str(&w, "room", room);
    if (password.len > 0) jw_kv_str(&w, "password", password);
    if (session.len > 0) jw_kv_str(&w, "session", session);
    return env_end(&w);
}

Str8 proto_encode_msg(Arena *a, const VsMsg *m) {
    JsonWriter w;
    jw_init(&w, a);
    switch (m->kind) {
        case VS_MSG_PING:
            env_begin(&w, "ping");
            jw_kv_i64(&w, "t", m->t);
            break;
        case VS_MSG_SET_READY:
            env_begin(&w, "setReady");
            jw_kv_bool(&w, "ready", m->ready);
            break;
        case VS_MSG_SET_FILE:
            env_begin(&w, "setFile");
            jw_kv_str(&w, "name", strbuf_str(&m->name));
            jw_kv_num(&w, "durationSec", m->duration_sec);
            jw_kv_i64(&w, "sizeBytes", m->size_bytes);
            break;
        case VS_MSG_CONTROL:
            env_begin(&w, "control");
            jw_key(&w, "action");
            jw_cstr(&w, vs_action_name(m->action));
            jw_kv_num(&w, "positionSec", m->position_sec);
            break;
        case VS_MSG_REPORT:
            env_begin(&w, "report");
            jw_kv_num(&w, "positionSec", m->position_sec);
            jw_kv_bool(&w, "paused", m->paused);
            jw_kv_bool(&w, "buffering", m->buffering);
            break;
        case VS_MSG_CHAT:
            env_begin(&w, "chat");
            jw_kv_str(&w, "text", strbuf_str(&m->text));
            break;
    }
    return env_end(&w);
}

void proto_hex(const u8 *bytes, isize n, char *out) { vs_hex_encode(bytes, n, out); }

b32 proto_session_token(char *out, isize cap) {
    if (cap < VS_SESSION_TOKEN_LEN + 1) return 0;
    u8 raw[VS_SESSION_TOKEN_BYTES];
    if (!vs_random_bytes(raw, (isize)sizeof(raw))) return 0;
    vs_hex_encode(raw, (isize)sizeof(raw), out);
    return 1;
}

// --------------------------------------------------------------- décodage ---

static void read_roomstate(const JsonValue *o, VsRoomState *rs) {
    memset(rs, 0, sizeof(*rs));
    rs->paused = json_get_bool(o, "paused", 0);
    rs->position_sec = json_get_num(o, "positionSec", 0);
    rs->rate = json_get_num(o, "rate", 0);
    rs->ref_server_ms = json_get_i64(o, "refServerMs", 0);
    strbuf_set(&rs->set_by, json_get_str(o, "setBy", str8_lit("")));
}

static isize read_users(Arena *a, const JsonValue *arr, VsUser **out) {
    *out = NULL;
    if (!arr || arr->kind != JSON_ARRAY || arr->count <= 0) return 0;
    isize n = arr->count;
    if (n > VS_MAX_USERS) n = VS_MAX_USERS;
    VsUser *users = arena_push_array(a, VsUser, n);
    isize i = 0;
    for (JsonValue *c = arr->first; c && i < n; c = c->next) {
        if (c->kind != JSON_OBJECT) continue;
        VsUser *u = &users[i++];
        u->id = json_get_str(c, "id", str8_lit(""));
        u->name = json_get_str(c, "name", str8_lit(""));
        u->ready = json_get_bool(c, "ready", 0);
        u->position_sec = json_get_num(c, "positionSec", 0);
        u->latency_ms = json_get_i64(c, "latencyMs", 0);
        JsonValue *f = json_get(c, "file");
        if (f && f->kind == JSON_OBJECT) {
            u->has_file = 1;
            u->file_name = json_get_str(f, "name", str8_lit(""));
            u->file_duration_sec = json_get_num(f, "durationSec", 0);
            u->file_size_bytes = json_get_i64(f, "sizeBytes", 0);
        }
    }
    *out = users;
    return i;
}

VsInMsg *proto_decode(Arena *a, Str8 raw) {
    JsonError err = JSON_OK;
    JsonValue *root = json_parse(a, raw, &err);
    if (!root || root->kind != JSON_OBJECT) return NULL;
    JsonValue *type = json_get(root, "type");
    if (!type || type->kind != JSON_STRING || type->string.len == 0) return NULL;
    JsonValue *data = json_get(root, "data");

    VsInMsg *m = arena_push_struct(a, VsInMsg);
    proto_fill(a, type->string, data, m);
    return m;
}

void proto_fill(Arena *a, Str8 type, const JsonValue *data, VsInMsg *m) {
    memset(m, 0, sizeof(*m));
    m->type = type;
    m->kind = VS_IN_UNKNOWN;

    if (str8_eq_cstr(m->type, "welcome")) {
        m->kind = VS_IN_WELCOME;
        m->self_id = json_get_str(data, "selfId", str8_lit(""));
        m->room = json_get_str(data, "room", str8_lit(""));
        JsonValue *st = json_get(data, "state");
        if (st && st->kind == JSON_OBJECT) {
            read_roomstate(st, &m->state);
            m->have_state = 1;
        }
        m->user_count = read_users(a, json_get(data, "users"), &m->users);
        for (isize i = 0; i < m->user_count; i++) {
            if (m->self_id.len > 0 && str8_eq(m->users[i].id, m->self_id)) {
                m->have_self_ready = 1;
                m->self_ready = m->users[i].ready;
                break;
            }
        }
    } else if (str8_eq_cstr(m->type, "pong")) {
        m->kind = VS_IN_PONG;
        m->pong.t = json_get_i64(data, "t", 0);
        m->pong.server_ms = json_get_i64(data, "serverMs", 0);
    } else if (str8_eq_cstr(m->type, "roomState")) {
        m->kind = VS_IN_ROOMSTATE;
        if (data && data->kind == JSON_OBJECT) {
            read_roomstate(data, &m->state);
            m->have_state = 1;
        }
    } else if (str8_eq_cstr(m->type, "users")) {
        m->kind = VS_IN_USERS;
        m->user_count = read_users(a, json_get(data, "users"), &m->users);
    } else if (str8_eq_cstr(m->type, "chatEvent") || str8_eq_cstr(m->type, "chat")) {
        m->kind = VS_IN_CHATEVENT;
        m->from = json_get_str(data, "from", str8_lit(""));
        m->text = json_get_str(data, "text", str8_lit(""));
        m->server_ms = json_get_i64(data, "serverMs", 0);
    } else if (str8_eq_cstr(m->type, "toast")) {
        m->kind = VS_IN_TOAST;
        m->level = json_get_str(data, "level", str8_lit("info"));
        m->text = json_get_str(data, "text", str8_lit(""));
    } else if (str8_eq_cstr(m->type, "error")) {
        m->kind = VS_IN_ERROR;
        m->code = json_get_str(data, "code", str8_lit(""));
        m->text = json_get_str(data, "text", str8_lit(""));
    }
}

b32 proto_error_is_fatal(Str8 code) {
    return str8_eq_cstr(code, "version_mismatch") || str8_eq_cstr(code, "bad_password") ||
           str8_eq_cstr(code, "name_taken");
}
